### PR TITLE
Add embeddings and search file support 

### PR DESCRIPTION
**Pull request by hallacy**
* Adds support for the embeddings endpoint
* Add search.prepare_data to the CLI to validate search files for upload

### CODE REVIEW

File: openai/api_resources/engine.py [hunk 1]
```diff
@@ -30,3 +30,6 @@ def generate(self, timeout=None, **params):
 
     def search(self, **params):
         return self.request("post", self.instance_url() + "/search", params)
+
+    def embeddings(self, **params):
```
Comment (madeleineth) [inline openai/api_resources/engine.py:34]: Can you take explicit parameters? I realize `**params` is common in this library but it's still bad.

File: openai/cli.py [hunk 2]
```diff
@@ -224,6 +227,41 @@ def list(cls, args):
 
 
 class Search:
+    @classmethod
+    def prepare_data(cls, args):
+
+        sys.stdout.write("Analyzing...\n")
```
Comment (madeleineth) [inline openai/cli.py:233]: Use `print`?

File: openai/cli.py [hunk 3]
```diff
@@ -224,6 +227,41 @@ def list(cls, args):
 
 
 class Search:
+    @classmethod
+    def prepare_data(cls, args):
```
Comment (madeleineth) [inline openai/cli.py:231]: Doc comment saying what this method does? It's confusing. I have no idea what the inputs or outputs are, or what it does, without reverse engineering.

File: openai/cli.py [hunk 4]
```diff
@@ -508,6 +511,29 @@ def help(args):
...
+        "-f",
+        "--file",
+        required=True,
+        help="JSONL, JSON, CSV, TSV, TXT or XLSX file containing prompt-completion examples to be analyzed."
+        "This should be the local file path.",
+    )
+    sub.add_argument(
+        "-p",
+        "--purpose",
+        help="Why are you uploading this file? (see https://beta.openai.com/docs/api-reference/ for purposes)",
```
Comment (madeleineth) [inline openai/cli.py:525]: Can you list the purposes? This specific thing has gotten me hung up in the past.

File: openai/cli.py [hunk 5]
```diff
@@ -508,6 +511,29 @@ def help(args):
...
+        required=True,
+    )
+    sub.add_argument(
+        "-q",
+        "--quiet",
+        required=False,
+        action="store_true",
+        help="Auto accepts all suggestions, without asking for user input. To be used within scripts.",
```
Comment (madeleineth) [inline openai/cli.py:533]: `-y`/`--yes` would be clearer, I think.

File: openai/validators.py [hunk 6]
```diff
@@ -70,7 +70,7 @@ def lower_case_column_creator(df):
     )
 
 
-def additional_column_validator(df):
+def additional_column_validator(df, fields=["prompt", "completion"]):
     """
     This validator will remove additional columns from the dataframe.
```
Comment (madeleineth) [inline openai/validators.py:75]: What is the meaning of `fields`? Either comment or give it a better name.

File: openai/validators.py [hunk 7]
```diff
@@ -101,50 +99,47 @@ def necessary_fn(x):
     )
 
 
-def non_empty_completion_validator(df):
+def non_empty_field_validator(df, field="completion"):
     """
     This validator will ensure that no completion is empty.
```
Comment (madeleineth) [inline openai/validators.py:104]: Same, for `field`.

File: openai/validators.py [hunk 8]
```diff
@@ -743,6 +738,30 @@ def write_out_file(df, fname, any_remediations, auto_accept):
...
+        fnames = get_outfnames(fname, split=False)
+
+        assert len(fnames) == 1
+        df[fields].to_json(fnames[0], lines=True, orient="records", force_ascii=False)
+
+        sys.stdout.write(
+            f'\nWrote modified file to {fnames[0]}`\nFeel free to take a look!\n\nNow upload that file:\n> openai api files.create -f "{fnames[0]}" -p {purpose}'
+        )
+    else:
+        sys.stdout.write("Aborting... did not write the file\n")
```
Comment (madeleineth) [inline openai/validators.py:762]: `print`

File: openai/validators.py [hunk 9]
```diff
@@ -799,3 +818,71 @@ def get_validators():
...
+        sys.stdout.write(
+            "\n\nBased on the analysis we will perform the following actions:\n"
+        )
+        for remediation in optional_remediations:
+            df, optional_applied = apply_optional_remediation(
+                df, remediation, auto_accept
+            )
+            any_optional_applied = any_optional_applied or optional_applied
+    else:
+        sys.stdout.write("\n\nNo remediations found.\n")
```
Comment (madeleineth) [inline openai/validators.py:884]: `print`

File: openai/cli.py [hunk 10]
```diff
@@ -508,6 +511,29 @@ def help(args):
     )
     sub.set_defaults(func=FineTune.prepare_data)
 
+    sub = subparsers.add_parser("search.prepare_data")
+    sub.add_argument(
+        "-f",
+        "--file",
+        required=True,
+        help="JSONL, JSON, CSV, TSV, TXT or XLSX file containing prompt-completion examples to be analyzed."
```
Comment (rachellim) [inline openai/cli.py:519]: Copy pasta of `prompt-completion` from fine tuning?

File: openai/cli.py (hunk 4 above)
Comment (rachellim) [inline openai/cli.py:525]: It seems unintuitive to me that the `Search` purpose (which is what this is ostensibly under) is further split into different purposes. Maybe each of them should be a different tool instead? (Or do they all fall under the search umbrella, and how? We can chat outside this PR if i'm missing some context)

**Comment by hallacy** [comment]
I think we're gonna split the validation code from the embeddings code. Closing this request down

**Pull request closed without merging**